docs: add evaluating traces notebook (#4521)
Arize-ai/phoenix: docs/evaluation/how-to-evals/evaluating-phoenix-traces.md

> # Overview

Reviewer: The title probably needs to match what shows up in the page header

> This guide will walk you through the process of evaluating traces captured in Phoenix, and exporting the results to the Phoenix UI.

Reviewer: ```suggestion
This guide will walk you through exporting traces captured in Phoenix, evaluating them programmatically, and  logging the evaluation results back to Phoenix for analysis.
```

> This process is similar to the [evaluation quickstart guide](https://docs.arize.com/phoenix/evaluation/evals), but instead of creating your own dataset or using an existing external one, you'll export a trace dataset from Phoenix and log the evaluation results to Phoenix.

Reviewer: ```suggestion
This builds on the [evaluation quickstart guide](https://docs.arize.com/phoenix/evaluation/evals), but instead of creating your own dataset or using an existing external one, you'll query traces from Phoenix to create a dataset and log the evaluation results to Phoenix so that the traces are annotated with the evaluations.
```

## Install dependencies & Set environment variables

```bash
%%bash
pip install -q "arize-phoenix>=4.29.0"
pip install -q openai
```

```python
import os
from getpass import getpass

import dotenv

dotenv.load_dotenv()

if not (openai_api_key := os.getenv("OPENAI_API_KEY")):
    openai_api_key = getpass("🔑 Enter your OpenAI API key: ")

os.environ["OPENAI_API_KEY"] = openai_api_key
```

## Connect to Phoenix

```python
# Check if PHOENIX_API_KEY is present in the environment variables.
# If it is, we'll use the cloud instance of Phoenix. If it's not, we'll start a local instance.
# A third option is to connect to a docker or locally hosted instance.
# See https://docs.arize.com/phoenix/setup/environments for more information.

import os

if "PHOENIX_API_KEY" in os.environ:
    os.environ["PHOENIX_CLIENT_HEADERS"] = f"api_key={os.environ['PHOENIX_API_KEY']}"
    os.environ["PHOENIX_COLLECTOR_ENDPOINT"] = "https://app.phoenix.arize.com"

>     print("Using cloud instance of Phoenix.")

Reviewer: this will get confusing soon

```bash
%%bash
pip install -q openinference-instrumentation-openai
```

```python
from openinference.instrumentation.openai import OpenAIInstrumentor

OpenAIInstrumentor().instrument(tracer_provider=tracer_provider)
```

```python
from openai import OpenAI

# Initialize OpenAI client
client = OpenAI()

# Function to generate a joke
def generate_joke():
    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "system", "content": "You are a helpful assistant that generates jokes."},
            {"role": "user", "content": "Tell me a joke."},
        ],
    )
    joke = response.choices[0].message.content
    return joke

# Generate 5 different jokes
jokes = []
for _ in range(5):
    joke = generate_joke()
    jokes.append(joke)
    print(f"Joke {len(jokes)}:\n{joke}\n")

print(f"Generated {len(jokes)} jokes and tracked them in Phoenix.")
```

## Download trace dataset from Phoenix

```python
import phoenix as px

spans_df = px.Client().get_spans_dataframe(project_name="evaluating_traces_quickstart")
spans_df.head()
```

## Generate evaluations

> Now that we have our trace dataset, we can generate evaluations for each trace. Evaluations can be generated in many different ways. Ultimately, we want to end up with a set of labels and/or scores for our traces.

Reviewer: ```suggestion
Now that we have our trace dataset, we can generate evaluations for each span. Evaluations can be generated in many different ways. Ultimately, we want to end up with a set of labels and/or scores for our traces.
```

You can generate evaluations using:
- Plain code

> - Phoenix's [built-in LLM as a Judge evaluators](https://docs.arize.com/phoenix/evaluation/how-to-evals/running-pre-tested-evals)

Reviewer: ```suggestion
- Phoenix's [built-in LLM evaluators](https://docs.arize.com/phoenix/evaluation/how-to-evals/running-pre-tested-evals)
```

> - Your own [custom LLM as a Judge evaluator](https://docs.arize.com/phoenix/evaluation/how-to-evals/bring-your-own-evaluator)

Reviewer: ```suggestion
- Your own [custom LLM  evaluator](https://docs.arize.com/phoenix/evaluation/how-to-evals/bring-your-own-evaluator)
```

- Other evaluation packages

As long as you format your evaluation results properly, you can upload them to Phoenix and visualize them in the UI.

Let's start with a simple example of generating evaluations using plain code. OpenAI has a habit of repeating jokes, so we'll generate evaluations to label whether a joke is a repeat of a previous joke.

```python
# Create a new DataFrame with selected columns
eval_df = spans_df[["context.span_id", "attributes.llm.output_messages"]].copy()
eval_df.set_index("context.span_id", inplace=True)

# Create a list to store unique jokes
unique_jokes = set()

# Function to check if a joke is a duplicate
def is_duplicate(joke_data):
    joke = joke_data[0]["message.content"]
    if joke in unique_jokes:
        return True
    else:
        unique_jokes.add(joke)

>         return False

Reviewer: I think this evaluation process might be slightly too hard to relate to -  I wonder if we can find a better "simple" eval
